Kitodo/src/main/java/org/kitodo/model/Subfolder.java

```diff
@@ -0,0 +1,301 @@
+
+/**
+ * A subfolder is a folder in the file system below the process folder that
+ * contains a variant of each file of the digital representation of the work
+ * intended for a particular use. In this approach, a file is a fragment of the
+ * digital representation of the artwork, a conceptual entity that can have
+ * different variants. The individual variants of this file are then each stored
+ * as computer-readable files in a folder for that use on the volume. They can
+ * have different formats, both in terms of the data type and their form of
+ * representation. The computer files, which are located in different use
+ * folders and which are conceptually the same file, share a part of their
+ * filename. This is referred to here as the canonical part of the file name.
+ * 
+ * <p>
+ * For historical reasons, in some institutions using this software, there is a
+ * situation where the different variants of a file are in the same folder and
+ * differ in filename suffix. We do not want to advertise this setup, but we
+ * will continue to support it because it would be difficult for these
+ * institutions to migrate the files.
+ */
+public class Subfolder {
+    /**
+     * The general metrics of this kind of subfolder. So to say its type.
+     */
+    private final Folder folder;
+
+    /**
+     * The process this subfolder belongs to. That is the process whose process
+     * directory this sub-folder is below on the file storage.
+     */
+    private final Process process;
+
+    /**
+     * The variable replacer used to replace variables in the folder path to the
+     * task folder.
+     */
+    private final VariableReplacer variableReplacer;
+
+    /**
+     * Convenience function to create a bunch of subfolders in one.
+     * 
+     * @param process
+     *            the process this subfolder belongs to
+     * @param folders
+     *            The general metrics of the kinds of subfolders to create
+     * @return subfolders
+     */
+    public static List<Subfolder> createAll(Process process, Collection<Folder> folders) {
+        return folders.parallelStream().map(lambda -> new Subfolder(process, lambda)).collect(Collectors.toList());
+    }
+
+    /**
+     * Creates a new subfolder.
+     * 
+     * @param process
+     *            the process this subfolder belongs to
+     * @param folder
+     *            the general metrics of this kind of subfolder
+     */
+    public Subfolder(Process process, Folder folder) {
+        this.process = process;
+        this.folder = folder;
+        this.variableReplacer = new VariableReplacer(null, null, process, null);
+    }
+
+    /**
+     * Returns a key mapper for a compiled regular expression.
+     *
+     * @param pattern
+     *            compiled regular expression to determine the key
+     * @return a mapping function for the key of the map
+     */
+    private static Function<URI, String> createKeyMapperForPattern(final Pattern pattern) {
```

Comment:
Sorry, I don't get, what is happening here. What is the key mapper used for? Why is it a Function?